will use queries plus evidence and voting

turbomam opened the pull request.

turbomam committed:
will use queries plus evidence and voting

turbomam commented:
Hi @sierra-moxon 

@mslarae13 started this architectural decision record about the environmental triad value set process. I updated it to include the stuff we reviewed yesterday, including the evidence tables.

I added you as a reviewer because I called you out for your future role in implementing my workflow as a Jupyter notebook or a single, more configurable application. Please add your feedback here, or the three of us can chat on Slack.

sierra-moxon commented:
I am happy to approve this PR as is. Thank you for writing it up!! ✅ 

I can imagine it being helpful to see the steps (and the code that we used to execute each of those steps) in some form of documentation (probably beyond the scope of the ADR specifically) so that if the technical person doing the next phase of this work (e.g. doing this exercise for "water"), has an SOP to recreate the process.  On advice from Chris, the Jupyter notebook mentioned in this ADR would be a good technical documentation entry point.  

The configurable yaml also mentioned here, will serve as the non-technical entry point for SMEs or other NMDC staff, to help refine the oak queries in an approachable way without diving into the meaning of the short-hand oak command or makefile syntax that is required to, for example, subtract specific terms, include terms, etc.  If they want to learn that syntax, that is terrific, but extracting it into a simple file format would make curation easier for folks with limited time.

I'd also advocate strongly that the code we use for this task (including jupyter notebook, all code required to do _just_ this env_triad PV task specifically, the resulting tables/files of PVs, and the visualizations, be checked in to their own _new_ repository. In my opinion, this modularization by repo will help non-squad members contribute to the maintenance of our code over time by narrowing the context they need to learn in order to participate, increasing control over the release process for this work separate from the release process for other scripts in the existing repository, and make this code "production ready" so that folks can reliably deploy new releases (if we get to that point) with no prior experience in the repo itself (as we expect to execute this code repeatedly (with likely tweaks) over time and with different environments/extensions/packages (e.g. water) and we want this code to be maintainable by folks outside the squad itself). 

I don't _think_ this feedback belongs in the ADR.

mslarae13 commented on decisions/0015-env-triad-terms.md:
```diff
@@ -1,52 +1,124 @@
-Determining what to assert for each of the environmental triad slot (`env_broad_scale`, `env_local_scale`, and `env_medium`) is difficult.
-NMDC will provide a reporducible logic to provide users with a curated lists of valid terms for each slot.
+Determining what values to populate into each of the environmental triad slots (`env_broad_scale`, `env_local_scale`,
+and `env_medium`) is difficult, especially when further broken down into reasonable `env_broad_scale` values for Soil
```
```suggestion
and `env_medium`) is difficult.
```

mslarae13 commented on decisions/0015-env-triad-terms.md:
```diff
@@ -1,52 +1,124 @@
-NMDC will provide a reporducible logic to provide users with a curated lists of valid terms for each slot.
+Determining what values to populate into each of the environmental triad slots (`env_broad_scale`, `env_local_scale`,
+and `env_medium`) is difficult, especially when further broken down into reasonable `env_broad_scale` values for Soil
+sample, Water samples, etc. Although it is challenging, it adds value to the NMDC ecosystem in terms of finding and
```
```suggestion

```

mslarae13 commented on decisions/0015-env-triad-terms.md:
```diff
@@ -1,52 +1,124 @@
+Determining what values to populate into each of the environmental triad slots (`env_broad_scale`, `env_local_scale`,
+and `env_medium`) is difficult, especially when further broken down into reasonable `env_broad_scale` values for Soil
+sample, Water samples, etc. Although it is challenging, it adds value to the NMDC ecosystem in terms of finding and
+grouping similar Biosamples.
```
```suggestion
```

I don't think we need this much extra validation for an ADR.

mslarae13 commented on decisions/0015-env-triad-terms.md:
```diff
@@ -1,52 +1,124 @@
+enumeration, or a regular expression pattern as constraints. This could be a temporary allowance while NMDC's logic is
+being refined. Values that match the regular expression but do not come from the enumeration will need to be reviewed on
+a regular basis, as submitters could provide values that have typos, that have CURIe/label mismatches, or just are not
+reasonable values for the slot.
```
This section is valuable text, but IDK if we need it in the ADR...

mslarae13 commented on decisions/0015-env-triad-terms.md:
```diff
@@ -1,52 +1,124 @@
+of the environments (MIxS Extensions), and will be reflected in either query-specific columns or general boolean
+columns, like 'is_biome' or 'is_environmental_material'.
+
+- The environmental triad squad is developing an NMDC-prioritized list of environments. Soil will be completed first.
```
```suggestion

```

not needed for the ADR

mslarae13 commented on decisions/0015-env-triad-terms.md:
```diff
@@ -1,52 +1,124 @@
+
+- The environmental triad squad is developing an NMDC-prioritized list of environments. Soil will be completed first.
+- This ADR will be updated as environment-specific queries are created.
+- The logic described above is intended to minimize cherry-picking of values for the sets, but individual squad members
```
```suggestion
- The logic described above is intended to minimize cherry-picking of values for the sets.
```

mslarae13 commented on decisions/0015-env-triad-terms.md:
```diff
@@ -1,52 +1,124 @@
+- The environmental triad squad is developing an NMDC-prioritized list of environments. Soil will be completed first.
+- This ADR will be updated as environment-specific queries are created.
+- The logic described above is intended to minimize cherry-picking of values for the sets, but individual squad members
+  will be given the opportunity to add their own `Y` or `N` votes for each value in the tables. It may be possible to
+  feed the insights from these votes back into reusable logic, but that might require complex OAK queries over
+  inconsistent annotations, or grouping of values within a semantic embedding space. For example, there may be evidence
+  that [farm [ENVO:00000078]](https://www.ebi.ac.uk/ols4/ontologies/envo/classes/http%253A%252F%252Fpurl.obolibrary.org%252Fobo%252FENVO_00000078),
+  [agricultural field [ENVO:00000114]](https://www.ebi.ac.uk/ols4/ontologies/envo/classes/http%253A%252F%252Fpurl.obolibrary.org%252Fobo%252FENVO_00000114)
+  and [banana plantation [ENVO:00000161]](https://www.ebi.ac.uk/ols4/ontologies/envo/classes/http%253A%252F%252Fpurl.obolibrary.org%252Fobo%252FENVO_00000161)
+  should all be included in the `env_local_scale` enumeration for Soil samples. It may be more expeditious for
+  a human to make the judgment call to include farm and exclude the others, compared to the effort required to automate
+  that. Visualizations will be provided to illustrate these cases. See the end of this ADR.
```
```suggestion
  Expert review will be done on an initial list of accepted terms. Insights from review will be fed back into reusable logic, which that might require complex OAK queries. over
  inconsistent annotations, or grouping of values within a semantic embedding space. For example, there may be evidence
  that [farm [ENVO:00000078]](https://www.ebi.ac.uk/ols4/ontologies/envo/classes/http%253A%252F%252Fpurl.obolibrary.org%252Fobo%252FENVO_00000078),
  [agricultural field [ENVO:00000114]](https://www.ebi.ac.uk/ols4/ontologies/envo/classes/http%253A%252F%252Fpurl.obolibrary.org%252Fobo%252FENVO_00000114)
  and [banana plantation [ENVO:00000161]](https://www.ebi.ac.uk/ols4/ontologies/envo/classes/http%253A%252F%252Fpurl.obolibrary.org%252Fobo%252FENVO_00000161)
```

Too much for a short ADR. we can add complex text to the final version if it's necessary.

mslarae13 commented on decisions/0015-env-triad-terms.md:
```diff
@@ -1,52 +1,124 @@
+  a human to make the judgment call to include farm and exclude the others, compared to the effort required to automate
+  that. Visualizations will be provided to illustrate these cases. See the end of this ADR.
+- As necessary, we can request that EnvO add classes in support of this work, or that more (and more consistent) axioms
+  are added to existing classes.
```
```suggestion
  are added to existing classes. This will not be done as part of the squad contributing this ADR.
```

mslarae13 commented on decisions/0015-env-triad-terms.md:
```diff
@@ -1,52 +1,124 @@
+- As necessary, we can request that EnvO add classes in support of this work, or that more (and more consistent) axioms
+  are added to existing classes.
 
-To ensure interoperability and consistancy, the NMDC General Query will be ammended for each environmental sample type.
```
I'm not sure why you deleted **all** of my text. But I would like it returned. It's clear and concise.

mslarae13 commented on decisions/0015-env-triad-terms.md:
```diff
@@ -1,52 +1,124 @@
-- `env_medium` will be composed of terms that branch from environmental material [ENVO:00010483]
+- `env_broad_scale` will consist of subclasses
+  of [biome [ENVO:00000428]](https://www.ebi.ac.uk/ols4/ontologies/envo/classes/http%253A%252F%252Fpurl.obolibrary.org%252Fobo%252FENVO_00000428),
+  with the exception of host-associated samples, including plant-associated samples.
```
```suggestion
  with the exception of host-associated samples, including plant-associated samples.
```

What? exception of host, ok that makes snes
but what plant-associated? IDK what you're saying here

mslarae13 commented on decisions/0015-env-triad-terms.md:
```diff
@@ -1,52 +1,124 @@
+  and [environmental material [ENVO:00010483]](https://www.ebi.ac.uk/ols4/ontologies/envo/classes/http%253A%252F%252Fpurl.obolibrary.org%252Fobo%252FENVO_00010483)
+    - It's expected that additional branches of the is_a hierarchy will subtracted,
+      like [chemical entity [CHEBI:24431]](https://www.ebi.ac.uk/ols4/ontologies/envo/classes/http%253A%252F%252Fpurl.obolibrary.org%252Fobo%252FCHEBI_24431)
+    - The possibility of subtracting leaf nodes is under consideration. Since we are programmatically building candidate
+      tables with multiple forms of evidence, the smaller query result offered by subtraction may not be of much added
+      benefit
```
```suggestion
    
```

remove possibilities. Just provide what we did. And IF we decide to do this AFTER manual review, we'll add it.

mslarae13 commented on decisions/0015-env-triad-terms.md:
```diff
@@ -1,52 +1,124 @@
+    - The possibility of subtracting leaf nodes is under consideration. Since we are programmatically building candidate
+      tables with multiple forms of evidence, the smaller query result offered by subtraction may not be of much added
+      benefit
+    - Subtraction over relationships other than is_a, subtraction of classes based on substrings in their labels, or
+      other targets subtraction are possible, but may also be of limited benefit
```
```suggestion

```

mslarae13 commented on decisions/0015-env-triad-terms.md:
```diff
@@ -1,52 +1,124 @@
+      benefit
+    - Subtraction over relationships other than is_a, subtraction of classes based on substrings in their labels, or
+      other targets subtraction are possible, but may also be of limited benefit
+    - @turbomam
+      has [prototyped these workflows in Makefiles](https://github.com/microbiomedata/external-metadata-awareness/blob/main/Makefiles/soil-env_local_scale.Makefile).
+      @sierra-moxon will explore options for making the workflows more accessible to the whole group, with a single
+      application that takes a single configuration file, or with (a) Jupyter notebook(s)
```
```suggestion
 
```

Not what the ADR is about. Exploration, possibilities, and options aren't a decision. Leave out until we decide.

mslarae13 commented on decisions/0015-env-triad-terms.md:
```diff
@@ -1,52 +1,124 @@
+
+- `env_broad_scale` will
+  exclude [aquatic biome [ENVO:00002030]](https://www.ebi.ac.uk/ols4/ontologies/envo/classes/http%253A%252F%252Fpurl.obolibrary.org%252Fobo%252FENVO_00002030)
+    - See comments about the cost and benefits of excluding leaf nodes above
```
```suggestion
```

mslarae13 commented on decisions/0015-env-triad-terms.md:
```diff
@@ -1,52 +1,124 @@
-- `env_medium` will be textually filtered for classes with soil or soils in their name or description.
-  - Will evaluate if adding stemming is needed.
+- `env_medium`
+    - Subclass hierarchy-based exclusion _could_ also be used here, for
```
```suggestion
    - Subclass hierarchy-based exclusion _could_ also be used here, for
```

Again, I don't want what COULD be used. I want to know what you did to create this inital list, pre manual review.

mslarae13 commented on decisions/0015-env-triad-terms.md:
```diff
@@ -1,52 +1,124 @@
+- `env_medium`
+    - Subclass hierarchy-based exclusion _could_ also be used here, for
+      example [hydrocarbon-based environmental material [ENVO:2000045]](https://www.ebi.ac.uk/ols4/ontologies/envo/classes/http%253A%252F%252Fpurl.obolibrary.org%252Fobo%252FENVO_2000045?lang=en)
+    - Textual filtering over labels, with or without regular expressions or stemming, _could_ also be used
+        - food, water and ice terms could be removed without stemming
+        - stemming with the pattern 'ferment' could be used to remove terms with fermented or fermenting in their labels
```
Ok, but did you? If not take it out. I asked you to add details about what you **did**

mslarae13 replied:
not could do.

mslarae13 commented on decisions/0015-env-triad-terms.md:
```diff
@@ -1,52 +1,124 @@
+    - Textual filtering over labels, with or without regular expressions or stemming, _could_ also be used
+        - food, water and ice terms could be removed without stemming
+        - stemming with the pattern 'ferment' could be used to remove terms with fermented or fermenting in their labels
+    - But there's a good chance that the empirical ranking of the environmental material classes will have a better
+      cost/benefit profile (at least initially) than writing a library of rules.
```
```suggestion

```

mslarae13 commented:
& @turbomam is there a reason you made another branch and PR and didn't just suggest changes on my PR like I did here,..?

turbomam commented:
> & @turbomam is there a reason you made another branch and PR and didn't just suggest changes on my PR like I did here,..?

Yes, good question.

Because I made edits on many lines and auto-formatted it to stay within ~ 120 characters per line, it **looks** like I deleted a lot of your work. But I really didn't. I think I followed your flow and overall message very closely. 

Did you read my version as a standalone document, as a diff view, or both? If you have time, would you be willing to read it as a free-standing document? It feel like it would have been hard for me to share my suggestions as comments, and I didn't want to make commits directly into your PR.

I agree 100% with your comments about not getting into "what could be done". I left that in there so you could see the continuity between what we have discussed over the past week or two. If you understand that I don't see a role for textual filtering with stemming any more, etc., and that's OK with you, then yes, we should delete that.

It's your PR so we can proceed however you want. If you want my help reconciling the two versions, I can help, but if you'd rather do it your self, that's OK with me. I think the only part I added that I would push for keeping is

> Determining what values to populate into each of the environmental triad slots (`env_broad_scale`, `env_local_scale`, and `env_medium`) is difficult, especially when further broken down into reasonable `env_broad_scale` values for `Soil` sample, `Water` samples, etc. Although it is challenging, it adds value to the NMDC ecosystem in terms of finding and grouping similar Biosamples.

In my opinion, making value sets for `env_broad_scale` etc on their own isn't that hard. It's the further breakdown by environment that has made this so challenging, but will also add the most value.

mslarae13 commented:
Hi @turbomam ! 

If you can respond to or accept the edits I made, I'm happy to merge this PR into my PR. I think the details you have about what could be done should be added to the notes doc or the discussion threads that exist (https://github.com/microbiomedata/issues/discussions) & when we're ready for the updated version of this ADR, we can add in these 'what can be done'  that was used.

turbomam commented:
I agree with all of your suggestions except that I think we should leave in the following, because it clarifies the most challenging part of the task and the greatest payoff.

> Determining what values to populate into each of the environmental triad slots (env_broad_scale, env_local_scale, and env_medium) is difficult, especially when further broken down into reasonable env_broad_scale values for Soil sample, Water samples, etc. Although it is challenging, it adds value to the NMDC ecosystem in terms of finding and grouping similar Biosamples.

mslarae13 committed:
Update decisions/0015-env-triad-terms.md

mslarae13 committed:
Update decisions/0015-env-triad-terms.md

mslarae13 committed:
Update decisions/0015-env-triad-terms.md

mslarae13 committed:
Update decisions/0015-env-triad-terms.md

mslarae13 committed:
Update decisions/0015-env-triad-terms.md

mslarae13 committed:
Update decisions/0015-env-triad-terms.md

mslarae13 committed:
Update decisions/0015-env-triad-terms.md

mslarae13 committed:
Update decisions/0015-env-triad-terms.md

mslarae13 committed:
Update decisions/0015-env-triad-terms.md

mslarae13 committed:
Update decisions/0015-env-triad-terms.md

mslarae13 commented on decisions/0015-env-triad-terms.md:
```diff
@@ -1,52 +1,108 @@
 
-*This ADR is in progress. Initialy plan has been outlined here & may change*
+*This ADR is in progress. The initial plan has been outlined here and may change.*
 
```
```suggestion
*This ADR will be updated as environment specific queries are created.
```

mslarae13 committed:
Update decisions/0015-env-triad-terms.md

mslarae13 commented on decisions/0015-env-triad-terms.md:
```diff
@@ -1,52 +1,108 @@
+reasonable values for the slot.
 
+*This ADR is in progress. The initial plan has been outlined here and may change.*
+*This ADR will be updated as environment specific queries are created.
```
```suggestion
```

mslarae13 committed:
Update decisions/0015-env-triad-terms.md
